x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/rolemapping/TransportClusterStateRoleMappingTranslator.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.action.rolemapping;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.xpack.core.security.authc.support.mapper.ExpressionRoleMapping;
+import org.elasticsearch.xpack.core.security.authc.support.mapper.ReservedRoleMappingXContentNameFieldHelper;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Translator class for ensuring unique API names and marking cluster-state role mappings as read-only.
+ * Role mappings retrieved from cluster-state are surfaced through both the transport and REST layers,
+ * along with native role mappings. Unlike native role mappings, cluster-state role mappings are
+ * read-only and cannot be modified via APIs. It is possible for cluster-state and native role mappings
+ * to have overlapping names.
+ *
+ * <p>
+ * This class handles the following responsibilities to ensure correct processing of cluster-state role mappings:
+ * </p>
+ *
+ * <ol>
+ *   <li>Appends a reserved suffix to cluster-state role mapping names to avoid conflicts with native role mappings.</li>
+ *   <li>Marks the metadata of cluster-state role mappings with a reserved read-only flag.</li>
+ *   <li>Removes internal metadata flags used in processing (see {@link ReservedRoleMappingXContentNameFieldHelper}).</li>
+ * </ol>
+ */
+public final class TransportClusterStateRoleMappingTranslator {
```

Comment:
I struggle with the naming of this class a little and I think it's because it does a lot of different things (translate, validate, mutate and inspect). 

I think these static methods belong in the actions where they're used and the constants belong in `ExpressionRoleMapping` since this is specific to the behaviour of that class.

Reply:
Agreed, I've dismantled this and moved its fields and methods, foremost to `ExpressionRoleMapping`